output/upload.go

```diff
@@ -2,100 +2,108 @@ package output
-	var err error
-
-	if !grant.Valid {
-		return "", fmt.Errorf("Error - can't upload without valid S3 grant")
+func SetupSnapshotUploadForAllServers(ctx context.Context, servers []*state.Server, opts state.CollectionOpts, logger *util.Logger) {
+	if opts.ForceEmptyGrant {
+		return
 	}
-
-	if grant.S3URL == "" && grant.LocalDir != "" {
-		location := grant.LocalDir + filename
-		err = os.MkdirAll(filepath.Dir(location), 0755)
-		if err != nil {
-			logger.PrintError("Error creating target directory: %s", err)
-			return "", err
-		}
-
-		err = ioutil.WriteFile(location, data.Bytes(), 0644)
-		if err != nil {
-			logger.PrintError("Error writing local file: %s", err)
-			return "", err
-		}
-		return location, nil
+	for idx := range servers {
+		go func(server *state.Server) {
+			var compactLogStats map[string]uint8
+			var compactLogTime time.Time
+			logger = logger.WithPrefixAndRememberErrors(server.Config.SectionName)
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case s := <-server.FullSnapshotUpload:
+					data, err := proto.Marshal(s)
+					if err != nil {
+						logger.PrintError("Error marshaling protocol buffers")
+						continue
+					}
+
+					err = uploadViaWebsocketOrHttp(ctx, server, logger, opts.TestRun, data, s.SnapshotUuid, s.CollectedAt.AsTime(), false)
+					if err != nil {
+						logger.PrintError("Error uploading snapshot: %s", err)
+					} else if !opts.TestRun {
+						logger.PrintInfo("Submitted full snapshot successfully")
+					}
+				case s := <-server.CompactSnapshotUpload:
+					data, err := proto.Marshal(s)
+					if err != nil {
+						logger.PrintError("Error marshaling protocol buffers")
+						continue
+					}
+
+					err = uploadViaWebsocketOrHttp(ctx, server, logger, opts.TestRun, data, s.SnapshotUuid, s.CollectedAt.AsTime(), false)
+					if err != nil {
+						logger.PrintError("Error uploading snapshot: %s", err)
+						continue
+					}
+					if opts.TestRun {
+						continue
+					}
+
+					kind := kindFromCompactSnapshot(s)
+					logger.PrintVerbose("Submitted compact %s snapshot successfully", kind)
+					if compactLogTime.IsZero() {
+						compactLogTime = time.Now().Truncate(time.Minute)
+						compactLogStats = make(map[string]uint8)
+						continue
+					}
+
+					compactLogStats[kind] = compactLogStats[kind] + 1
+					if time.Since(compactLogTime) > time.Minute {
+						var keys []string
+						for k := range compactLogStats {
+							keys = append(keys, k)
+						}
+						sort.Strings(keys)
+						details := ""
+						for i, kind := range keys {
+							details += fmt.Sprintf("%d %s", compactLogStats[kind], kind)
+							if i < len(keys)-1 {
+								details += ", "
+							}
+						}
+						if len(details) > 0 {
+							logger.PrintInfo("Submitted compact snapshots successfully: " + details)
+						}
+						compactLogTime = time.Now().Truncate(time.Minute)
+						compactLogStats = make(map[string]uint8)
+					}
```

Comment:
Would it make sense to extract this to a helper? It's a lot of code inline for just logging.